xtevent/_eventgenvars.ado

```diff
@@ -1,306 +1,499 @@
+				exit 301
+			}
+		}
+		
+		* Check for vars named _k
+		cap unab oldkvars : _k*
+		if !_rc {
+			di as err _n "You have variable names with the _k prefix. _k is reserved for internal -xtevent- variables."
+			di as err _n "Please rename these variables before proceeding."
+			exit 110
+		}
+	}
+
+	********************* find first and last observed values *********************
+
+	*find minimum valid time (time where there is a no-missing observation)
+	tempvar zmint zmint2 zminv zminv2 zmaxt zmaxt2 zmaxv zmaxv2
+	cap by `panelvar' (`timevar'): egen `zmint'=min(`timevar') if !missing(`z') & `touse'
+	cap by `panelvar' (`timevar'): egen `zmint2'=min(`zmint')
+	*find the corresponding minimum valid value
+	cap by `panelvar' (`timevar'): gen `zminv'=`z' if `timevar'==`zmint2' 
+	cap by `panelvar' (`timevar'): egen `zminv2'=min(`zminv')
+
+	*find maximum valid time
+	cap by `panelvar' (`timevar'): egen `zmaxt'=max(`timevar') if !missing(`z') & `touse'
+	cap by `panelvar' (`timevar'): egen `zmaxt2'=max(`zmaxt')
+	*find the corresponding maximum valid value
+	cap by `panelvar' (`timevar'): gen `zmaxv'=`z' if `timevar'==`zmaxt2' 
+	cap by `panelvar' (`timevar'): egen `zmaxv2'=max(`zmaxv')
+	
+	*create a copy of z. If imputation happens, it will be on this copy
+	cap drop zn2 
+	qui gen zn2=`z'
+
+	********** show a warning message if we don't know treatment time for some units due to missing values in policyvar 
+	tempvar zwd zwu 
+	qui gen `zwd'=`z' if `touse'
+	cap by `panelvar' (`timevar'): replace `zwd'=`zwd'[_n-1] if missing(`z') & `timevar'>=`zmint2' & `timevar'<=`zmaxt2' & `touse'
+	qui gen `zwu'=`z' if `touse'
+	gsort + `panelvar' - `timevar'
+	cap by `panelvar': replace `zwu'=`zwu'[_n-1] if missing(`z') & `timevar'>=`zmint2' & `timevar'<=`zmaxt2' & `touse'
+	sort `panelvar' `timevar'
+	cap assert `zwd'==`zwu' if missing(`z') & `timevar'>=`zmint2' & `timevar'<=`zmaxt2' & `touse'
+	if _rc{
+		di "Event time is unknown for some units due to missing values in policyvar."
+	}
+	
+	********** Verify consistency with staggered adoption *******************
+	
+	loc bin 0
+	if "`impute'"=="stag" | "`impute'"=="instag" {
+		tempvar zn l1   
+		****** Check if z is binary
+		cap assert inlist(`z',0,1,.) if `touse'
+		if _rc {
+			qui su `z' if `touse'
+			loc rminz=`=r(min)'
+			loc rmaxz=`=r(max)'
+			cap assert inlist(`z',`rminz',`rmaxz',.)
+			if !_rc {
+				if "`impute'"=="stag" | "`impute'"=="instag"{
+					di "Policyvar is binary, but its values are different from 0 and 1. Assuming `=r(min)' as the unadopted policy state and `=r(max)' as the adopted policy state."
+				}
+				loc bin 1
+			}
+			else loc bin 0
+		}
+		else {
+			loc rminz=0
+			loc rmaxz=1
+			loc bin 1
+		}
+	}
+
+	* If not binary, default 
+	if `bin'==0 & ("`impute'"=="stag" | "`impute'"=="instag") {
+		
+		di "The policy variable is not binary. Assuming non-staggered adoption (no imputation)."
+		di "If event dummies and variables are saved, event-time will be missing."	

```

Comment:
If the user specified the option `savek( )` the event-time dummies and the event-time variable are added to the database, but the generation of the event-time variable is conditioned to the policy variable passing the staggered conditions, otherwise it is generated with missing values.

Reply:
Marking as resolved @Constantino-Carreto-Romero  @veli-m-andirin